Singleplayer/Multiplayer (RK3+RK4) implementation

a7emenov opened the pull request.

a7emenov committed:
Added temp rating change method (insecure)
Polished tests
Slightly edited application controller

a7emenov committed:
Added RatingChangeData class

a7emenov committed:
Singleplayer initial start

a7emenov committed:
Added flyway support
Slightly changed classes

a7emenov committed:
Updated migrations

a7emenov committed:
Updated Travis file and migrations

a7emenov committed:
Updated Travis file

a7emenov committed:
Updated Travis file

a7emenov committed:
Updated beta single-player methods
Slightly refactored code

a7emenov committed:
Bug correction

a7emenov committed:
Schema corrections
Added single game handling
Refactored code
Corrected some database errors

a7emenov committed:
Added manual exit method

a7emenov committed:
Added data
Slight fixes

a7emenov committed:
Temporary commit for testing

a7emenov committed:
Implemented sockets for singleplayer
Removed old code
Refactored code

a7emenov committed:
Completely remade sockets
Major code refactoring
Intermediate changes

a7emenov committed:
Socket additions
Work in progress

a7emenov committed:
Most work done

a7emenov committed:
Deleted temp stuff

a7emenov committed:
Added sending points

a7emenov committed:
Added approved origins

a7emenov committed:
Added exiting game
Fixed errors

a7emenov committed:
Added sending properties

a7emenov committed:
Added sending answers to all participants

a7emenov committed:
Fixed errors
Added correct error handling

a7emenov committed:
Temporary removed file

a7emenov committed:
Added correct answer resending
Small fixes

a7emenov committed:
Multiplayer game time fix

KSolovyev commented on src/main/java/websocket/GameManagerService.java:
```diff
@@ -0,0 +1,619 @@
+        checkQueue();
+    }
+
+    private void checkQueue() throws IOException {
```
synchronized

KSolovyev commented on src/main/java/websocket/GameManagerService.java:
```diff
@@ -0,0 +1,619 @@
+    private @NotNull WebSocketMessage<BaseGameContent> getGameState(
+        @NotNull ScheduledGame scheduledGame,
+        @NotNull String login,
+        boolean starting) throws IOException {
```
`boolean starting` выглядит не очень

a7emenov replied:
Немного изменил структуру метода, так что boolean starting более лаконично смотрится, но сам параметр необходим для того, чтобы правильно передать все фронту.

KSolovyev replied:
я бы сказал, что логичнее тогда вообще отдельный метод написать

KSolovyev commented on src/main/java/websocket/GameManagerService.java:
```diff
@@ -0,0 +1,619 @@
+    public static final int MULTIPLAYER_TIME_LIMIT = 120;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(GameManagerService.class);
+    private static final ScheduledExecutorService SCHEDULER = Executors.newSingleThreadScheduledExecutor();
```
вот это надо сделать обычным полем, не статическим

KSolovyev reviewed (changes requested):
Вообще код хороший, но сложный. Очень много условий `if(game.type == SINGLE_PLAYER) {...`Чувствуется, что в таких местах не хватает какой-нибудь фабрики, передаваемой при создании игры, которая смогла бы выполнять все эти действия без лишних условий. Я вам не советую сейчас что-то начинать менять, но просто прикинуть, как бы вышел код, можно

a7emenov committed:
Added multithreading

a7emenov committed:
Added queuemanager class
Small fixes

a7emenov committed:
Added joining in the middle of game

a7emenov committed:
Minor code changes

a7emenov committed:
QueueManager fix

a7emenov committed:
QueueManager fix2

a7emenov commented:
Отдельные методы на старт и просто получение состояния?

a7emenov committed:
Separated gameStateMessages

a7emenov committed:
Added getting random dashes for offline gaming

a7emenov committed:
Major code refactoring
Extracted subclasses of GameManagerService to different classes
Fixed minor bugs with exiting the game
Small sockethandler tweaks

a7emenov committed:
Fixed invalid winner when game stopping on timeout

a7emenov committed:
Fixed invalid winner when game stopping on timeout

a7emenov committed:
Added voting for answers, connecting/disconnecting from game messages
Minor code refactoring

a7emenov committed:
Deleted singleplayer/multiplayer game services
Updated database schema
Small fixes

a7emenov committed:
Added points to multiplayer game content
Scheduled game managers code refactoring

a7emenov committed:
Bugfix

a7emenov committed:
Small adjustments

a7emenov committed:
Fixed game not shutting down when painter leaves
Fixed id distribution when joining the game

KSolovyev commented:
хочется замержить. Вы все поправили?

a7emenov commented:
Да, давайте.
